splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+        return {
+            # should be overriden on startup to reflect sensible values
+            b'ttyname': (OptionType.fake, b'OK'),
+            b'ttytype': (OptionType.fake, b'OK'),
+            b'display': (OptionType.override, b':0'),
+            b'lc-ctype': (OptionType.fake, b'OK'),
+            b'lc-messages': (OptionType.fake, b'OK'),
+            b'putenv': (OptionType.fake, b'OK'),
+            b'allow-pinentry-notify': (OptionType.verify, None),
+            b'agent-awareness': (OptionType.verify, b'2.1.0')
+        }
+
+    @staticmethod
+    def default_timer_delay():
+        return {
+            'PKSIGN': None,     # always query for signing
+            'PKDECRYPT': 300    # 5 min
+        }
+
+    @staticmethod
+    def default_hash_algos():
+        return {
+            2: HashAlgo('sha1', 40),
+            3: HashAlgo('rmd160', 40),
+            8: HashAlgo('sha256', 64),
+            9: HashAlgo('sha384', 96),
+            10: HashAlgo('sha512', 128),
+            11: HashAlgo('sha224', 56),
+        }
+
+    @staticmethod
+    def notify(msg):
+        # TODO: call into dbus directly
+        subprocess.call(['notify-send', 'split-gpg2: {}'.format(msg)])
+
+    def request_timer(self, name):
+        now = time.time()
+        delay = self.timer_delay[name]
+        timestamp_path = self.timestamp_path(name)
+        if delay is not None:
+            try:
+                mtime = timestamp_path.stat().st_mtime
+                if mtime + delay > now:
+                    self.notify('command {} automatically allowed'.format(name))
+                    return
+            except FileNotFoundError:
+                pass
+
+        short_msg = "split-gpg2: '{}' wants to execute {}".format(
+            self.client_domain, name)
+        question = '{}\nDo you want to allow this{}?'.format(
+            short_msg,
+            'for the next {}s'.format(delay) if delay is not None else '')
+        if subprocess.call(['zenity', '--question', '--title', short_msg,
+                            '--text', question]) != 0:
+            raise Filtered
+
+        self.notify('command {} allowed'.format(name))
+        timestamp_path.touch()
+
+    def timestamp_path(self, name) -> pathlib.Path:
+        return pathlib.Path('{}_qubes-split-gpg2-timestamp_{}'.format(
+            self.agent_socket_path, name))
+
+    def client_write(self, data):
+        self.log_io('C <<<', data)
+        self.client_transport.write(data)
+
+    def eof_received(self) -> Optional[bool]:
+        # close connection to the real gpg agent too
+        if self.agent_transport is not None:
+            self.agent_transport.write_eof()
+        # automatically close the transport
+        return False
+
+    def data_received(self, untrusted_data: bytes) -> None:
+        # pylint: disable=arguments-differ
+        if len(self.untrusted_input_buffer) > ASSUAN_LINELENGTH*2:
+            self.log.error('Too much data received, dropping')
```

Comment:
You don't actually drop it? Also why `ASSUAN_LINELENGTH*2`?

I propose to not check length here but below (see comment). Before splitting the worst case should be an allocation error which should be fine.

Reply:
ok